README.md
# ANTsX evaluation in the [BraTS-Reg22 challenge](https://www.med.upenn.edu/cbica/brats-reg-challenge/)

* [Training phase leaderboard](https://www.cbica.upenn.edu/BraTSReg2022/lboardTraining.html)
* [Validation phase leaderboard](https://www.cbica.upenn.edu/BraTSReg2022/lboardValidation.html)

## Modalities to test on training data

* T1
* T1-contrast enhanced
* T2
* FLAIR
* T1-contrast enhanced + T2
* T1-contrast enhanced + FLAIR

## Parameter sets to test on training data

* antsRegistrationSyNQuick[a]
    * Rigid stage 
        * similarity metric:  MI with 32 bins
        * gradient step = 0.1
        * iterations per level = [1000,500,250,0]
        * shrink factors per level = [8,4,2,1]
        * smoothing factor per level = [3,2,1,0] (voxels)
    * Affine stage
        * same parameters as affine stage
* antsRegistrationSyNQuick[s,32]
    * Rigid stage 
        * Same as "antsRegistrationSyNQuick[a]"
    * Affine stage
        * Same as "antsRegistrationSyNQuick[a]"
    * SyN stage
        * Gaussian-based
        * similarity metric:  MI with 32 bins
        * gradient step = 0.1
        * iterations per level = [100,70,50,0]
        * shrink factors per level = [8,4,2,1]
        * smoothing factor per level = [3,2,1,0] (voxels) 
* antsRegistrationSyNQuick[b,32]
    * B-spline-based (mesh size per stage = [26,13,6.5,3.25] (mm)) 
    * other metrics same as "antsRegistrationSyNQuick[s,32]" 
* antsRegistrationSyN[s,2]
    * Rigid stage 
        * similarity metric:  MI with 32 bins
        * gradient step = 0.1
        * iterations per level = [1000,500,250,100]
        * shrink factors per level = [8,4,2,1]
        * smoothing factor per level = [3,2,1,0] (voxels)    
    * Affine stage
        * Same as rigid stage
    * SyN stage
        * Gaussian-based
        * similarity metric: CC with neighborhood radius of 2
        * gradient step = 0.1
        * iterations per level = [100,70,50,20]
        * shrink factors per level = [8,4,2,1]
        * smoothing factor per level = [3,2,1,0] (voxels)
* antsRegistrationSyN[b,2,26]
    * B-spline-based version of "antsRegistrationSyN[s,2]"
* antsRegistrationSyN[s,4]
    * Same as "antsRegistrationSyN[s,2]" with CC radius of 4
* antsRegistrationSyN[b,4,26]
    * B-spline-based version of "antsRegistrationSyN[s,4]"

## Training data results

### Accuracy

$$ \text{Accuracy improvement} = 1.0 - \frac{LandmarkError_{post}}{LandmarkError_{pre}} $$

<p align="middle">
  <img src="https://github.com/ntustison/BraTS-Reg22/blob/main/Analysis/accuracyBox.png" width="400" />
  <img src="https://github.com/ntustison/BraTS-Reg22/blob/main/Analysis/accuracy.png" width="400" /> 
</p>


### Timing (in seconds)

* Mac Pro 2020
* 2.5 GHz 28-Core Intel Xeon W
* `os.environ["ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS"] = "56"`

| Parameter set | Single modality | Two modalities |
| :---        |    :----:   |          ---: |
antsRegistrationSyNQuick[a]         |  4.6  |   NA
antsRegistrationSyNQuick[b,32,26]   | 39.8  |   38.5
antsRegistrationSyNQuick[s,32]      | 27.7  |  28.3
antsRegistrationSyN[s,2]            | 169.2 |  272.1
antsRegistrationSyN[b,2,26]         | 273.1 |  382.0
antsRegistrationSyN[s,4]            | 344.2 |  610.7
antsRegistrationSyN[b,4,26]         | 446.2 |  720.5

--------------------------------------------------------------------------------------------------------

## Apply antsRegistrationSyN[s,2] with T1-contrast enhanced to validation data

The choice of antsRegistrationSyN[s,2] with T1-contrast enhanced is due to combined 
performance and timing considerations.        
        
```python
>>> import ants
>>> import pandas as pd
>>>
>>> #######################################################################
>>> # I am making assumptions about file names.  Please change accordingly.
>>> 
>>> baseline_image = ants.image_read("baseline_t1ce.nii.gz")
>>> followup_image = ants.image_read("followup_t1ce.nii.gz")
>>> followup_landmark_file = "followup_landmarks.csv"
>>> warped_followup_landmark_file = "warped_followup_landmarks.csv"
>>>
>>> ########################################################################
>>> 
>>> reg = ants.registration(baseline_image, followup_image, type_of_transform="antsRegistrationSyN[s,2]", verbose=1)
>>>
>>> moving_indices = pd.read_csv(followup_landmark_file).drop('Landmark', axis=1)
>>> moving_indices = moving_indices.rename(columns={'X' : 'x', 'Y' : 'y', 'Z' : 'z'})
>>> moving_indices['y'] = moving_indices['y'] + 239
>>>
>>> moving_points = np.zeros(moving_indices.shape)
>>> for j in range(moving_indices.shape[0]):
>>>     moving_points[j,:] = ants.transform_index_to_physical_point(baseline_image, (moving_indices.iloc[j].values).astype(int))
>>> 
>>> moving_points_df = pd.DataFrame(data = {'x': moving_points[:,0], 'y': moving_points[:,1], 'z': moving_points[:,2]})
>>> moving_warped_points = ants.apply_transforms_to_points(3, moving_points_df, reg['invtransforms'], whichtoinvert=(True, False))   
>>> moving_warped_points = moving_warped_points.to_numpy()   
>>> moving_warped_points_df = pd.DataFrame(data=moving_warped_points, columns=['X', 'Y', 'Z'])
>>> moving_warped_points_df.insert(0, "Landmark", list(range(1, moving_points.shape[0]+1)))
>>> moving_warped_points_df.to_csv(warped_followup_landmark_file, index=False)
```
-----------------------------------------------------------

## Goodhart's Law

![GoodhartsLaw](https://user-images.githubusercontent.com/324811/204690199-685a44e3-b7d4-4d25-a19b-40b06c50c29d.png)






BraTSReg_Training_Data_v3/ChangeLog.txt
This is version 3 of the training dataset for Brain Tumor Sequence Registration (BraTS-Reg) challenge, 2022.

We identified 9 cases of the training set that required some updates in a few of the provided landmark annotations. These cases (indicated below) are now updated to more precise locations. 
BraTSReg_054
BraTSReg_062
BraTSReg_064
BraTSReg_076
BraTSReg_086
BraTSReg_127
BraTSReg_130
BraTSReg_135
BraTSReg_139

Analysis/plotResults.R
library( ggplot2 )
library( reshape )

baseDirectory <- "/Users/ntustison/Data/Public/BRATS/RegistrationCompetition2022/Analysis/"

resultsAffine <- read.csv( paste0( baseDirectory, "resultsAffine.csv" ) )
resultsSyNQuick <- read.csv( paste0( baseDirectory, "resultsSyNQuick.csv" ) )
resultsUniSyN <- read.csv( paste0( baseDirectory, "resultsUnivariateSyN.csv" ) )
resultsMultiSyN <- read.csv( paste0( baseDirectory, "resultsMultivariateSyN.csv" ) )
results <- rbind( resultsAffine, resultsSyNQuick, resultsUniSyN, resultsMultiSyN )


# transformTypes <- unique( resultsUniSyN$RegistrationType )
# modalities <- unique( resultsUniSyN$Modalities )
# subjects <- unique( resultsUniSyN$Subject )
# for( i in seq.int( length( subjects ) ) )
#   {
#   for( j in seq.int( length( transformTypes ) ) )
#     {
#     for( k in seq.int( length( modalities ) ) )
#       {
#       numberOfEntries <- length( which( resultsUniSyN$Subject == subjects[i] &
#                                         resultsUniSyN$RegistrationType == transformTypes[j] & 
#                                         resultsUniSyN$Modalities == modalities[k] ) )
#       if( numberOfEntries != 1 )
#          {
#          cat( subjects[i], ", ", transformTypes[j], ", ", modalities[k], ": ", numberOfEntries, "\n")
#          }
#       } 
#     }
#   }

results$Accuracy <- 1.0 - results$PostLandmarkError / results$PreLandmarkError

transformTypes <- unique( results$RegistrationType )
modalities <- unique( results$Modalities )

performanceMap <- array( data = NA, dim = c( length( transformTypes ), length( modalities ) ) )
rownames( performanceMap ) <- transformTypes
colnames( performanceMap ) <- modalities

timing <- rep( 0, length( transformTypes ) )
timingSingleModality <- rep( 0, length( transformTypes ) )
timingTwoModalities <- rep( 0, length( transformTypes ) )

for( i in seq.int( length( transformTypes ) ) ) 
  {
  timing[i] <- mean( results$TimeElapsed[which( results$RegistrationType == transformTypes[i] )])
  timingSingleModality[i] <- mean( results$TimeElapsed[which( results$RegistrationType == transformTypes[i] & 
                                                              results$Modalities %in% c( "t1", "t1ce", "flair", "t2" ) )])
  timingTwoModalities[i] <- mean( results$TimeElapsed[which( results$RegistrationType == transformTypes[i] & 
                                                              results$Modalities %in% c( "t1ce_t2", "t1ce_flair" ) )])
  for( j in seq.int( length( modalities ) ) )
    {
    # performanceMap[i, j] <- mean( results$Accuracy[which( results$RegistrationType == transformTypes[i] & results$Modalities == modalities[j] )])
    performanceMap[i, j] <- median( results$Accuracy[which( results$RegistrationType == transformTypes[i] & results$Modalities == modalities[j] )])
    }
  }
  
performanceMelt <- melt( performanceMap )
colnames( performanceMelt ) <- c( "TransformType", "Modality", "AccuracyImprovement" )
performanceMelt$TransformType <- factor( performanceMelt$TransformType, 
    levels = c( "antsRegistrationSyNQuick[a]", 
                "antsRegistrationSyNQuick[s,32]", "antsRegistrationSyNQuick[b,32,26]",                 
                "antsRegistrationSyN[s,4]", "antsRegistrationSyN[b,4,26]", 
                "antsRegistrationSyN[s,2]", "antsRegistrationSyN[b,2,26]", "brats" )  
    )
performanceMelt$Modality <- factor( performanceMelt$Modality, 
    levels = c( "t1", "t1ce", "flair", "t2", "t1ce_flair", "t1ce_t2" ) )

heatMapPlot <- ggplot( data = performanceMelt, aes( x = TransformType, y = Modality ) ) + 
               geom_tile( aes( fill = AccuracyImprovement ) ) + 
               geom_text( aes( label = round( AccuracyImprovement, 2 ) ) ) +
               theme( axis.text.x = element_text( angle = 90, hjust = 1, vjust = 0.5 ) )
ggsave( paste0( baseDirectory, "/accuracy.pdf" ), heatMapPlot, width = 7, height = 5 )

results$RegistrationType <- factor( results$RegistrationType, 
    levels = c( "antsRegistrationSyNQuick[a]", 
                "antsRegistrationSyNQuick[s,32]", "antsRegistrationSyNQuick[b,32,26]",                 
                "antsRegistrationSyN[s,4]", "antsRegistrationSyN[b,4,26]", 
                "antsRegistrationSyN[s,2]", "antsRegistrationSyN[b,2,26]", "brats" )  
    )
barPlot <- ggplot( data = results )   + 
           geom_boxplot( aes( x = RegistrationType, y = Accuracy, fill = Modalities ) ) +
           theme( axis.text.x = element_text( angle = 90, hjust = 1, vjust = 0.5 ) )
           
ggsave( paste0( baseDirectory, "/accuracyBox.pdf" ), barPlot, width = 7, height = 5 )


Scripts/organize_validation_data_for_upload.py
import os
import glob
import shutil

os.environ["ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS"] = "4"

base_directory = "/Users/ntustison/Data/Public/BRATS/RegistrationCompetition2022/"

subjects = glob.glob(base_directory + "Data/BraTSReg_Validation_Processed/BraTSReg_*")

for i in range(len(subjects)):
    
    print("\n" + subjects[i], "(", i+1, "out of", len(subjects), ")")

    subject_prefix = os.path.basename(subjects[i])

    subject_files = glob.glob(subjects[i] + "/*.csv")    
    for j in range(len(subject_files)):
        print("   " + subject_files[j]) 

        subject_file_basename = os.path.basename(subject_files[j])
        tokens = subject_file_basename.split("_")

        if "t1ce_flair" in subject_file_basename or "t1ce_t2" in subject_file_basename:
            if tokens[3] == 'b':
                experiment = "_".join(tokens[2:8])
            else:
                experiment = "_".join(tokens[2:7])
        else:
            if tokens[3] == 'b':
                experiment = "_".join(tokens[2:7])
            else:
                experiment = "_".join(tokens[2:6])

        output_directory = base_directory + "Data/BraTSReg_Validation_Processed_Upload" + "/" + experiment
        if not os.path.exists(output_directory):
            os.makedirs(output_directory, exist_ok=True)

        output_file = output_directory + "/" + subject_prefix
        if ".csv" in tokens[-1]:
            output_file += ".csv"
        else:
            continue

        # elif "moving_warped.nii.gz" in tokens[-1]:
        #     output_file += "Warped.nii.gz"
        # else:
        #     output_file += tokens[-1]

        if not os.path.exists(output_file):
            print(subject_files[j] + " --> " + output_file)        
            shutil.copy(subject_files[j], output_file)


Scripts/process_data.py
import ants
import numpy as np
import os
import glob
import pandas as pd
import math
import time

os.environ["ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS"] = "56"

base_directory = "/Users/ntustison/Data/Public/BRATS/RegistrationCompetition2022/Data/"

write_xfrm_output = True

multi_modalities = [
                    ['t1'], 
                    ['t1ce'],
                    ['flair'],
                    ['t2'],
                    ['t1ce', 'flair'],
                    ['t1ce', 't2']
                   ] 
single_modality = [
                    ['t1'], 
                    ['t1ce'],
                    ['flair'],
                    ['t2']
                   ] 
transform_types = [
                   "antsRegistrationSyNQuick[a]",
                    "antsRegistrationSyNQuick[s,32]",
                    "antsRegistrationSyNQuick[b,32,26]",
                    "antsRegistrationSyN[s,2]",
                    "antsRegistrationSyN[b,2,26]",
                    "antsRegistrationSyN[s,4]",
                    "antsRegistrationSyN[b,4,26]"
                  ]

results_df = pd.DataFrame(columns=["RegistrationType", "Modalities", "Subject", "TimeElapsed", "PreLandmarkError", "PostLandmarkError"])

subjects = glob.glob(base_directory + "BraTSReg_Validation_Data/BraTSReg_*")
# subjects = glob.glob(base_directory + "BraTSReg_Training_Data_v3/BraTSReg_*")

for i in range(len(subjects)):
    
    print("\n" + subjects[i], "(", i+1, "out of", len(subjects), ")")

    subject_prefix = os.path.basename(subjects[i])

    for t in range(len(transform_types)):

        type_of_transform = transform_types[t]
        
        if type_of_transform == "brats" or type_of_transform == "antsRegistrationSyNQuick[a]":
            which_modalities = single_modality
        else:
            which_modalities = multi_modalities

        for m in range(len(which_modalities)):

            modalities = which_modalities[m]          

            output_prefix = ""
            if write_xfrm_output:
                output_directory = (os.path.dirname(subjects[i]) + "/" + subject_prefix + "/").replace("BraTSReg_Validation_Data", "BraTSReg_Validation_Processed")
                if not os.path.exists(output_directory):
                    os.makedirs(output_directory, exist_ok=True)
                output_transform_string = type_of_transform.replace("[", "_")
                output_transform_string = output_transform_string.replace("]", "_")
                output_transform_string = output_transform_string.replace(",", "_")
                output_prefix = output_directory + subject_prefix + "_" + output_transform_string + '_'.join(modalities) + "_"

            print(output_prefix)
        
            if os.path.exists(output_prefix + "moving_warped_landmarks.csv"):
               continue

            baseline_images = list()
            followup_images = list()
            for j in range(len(modalities)):
                baseline_image_filename = subjects[i] + "/" + subject_prefix + "_00_0000_" + modalities[j] + ".nii.gz"
                if not os.path.exists(baseline_image_filename):
                    raise ValueError(baseline_image_filename + " does not exist") 
                baseline_images.append(ants.image_read(baseline_image_filename))
                
                followup_image_filenames = glob.glob(subjects[i] + "/" + subject_prefix + "_01_*" + modalities[j] + ".nii.gz")
                followup_images.append(ants.image_read(followup_image_filenames[0]))

                print("    " + modalities[j] + ": ")
                print("        " + baseline_image_filename)
                print("        " + followup_image_filenames[0])

            time_start = time.time()    
            regInvXfrms = None
            regFwdXfrms = None
            if len(baseline_images) == 1 or type_of_transform == "brats":
                reg = ants.registration(baseline_images[0], followup_images[0], type_of_transform=type_of_transform, outprefix=output_prefix, verbose=1)
                if type_of_transform == "antsRegistrationSyNQuick[a]":
                    regInvXfrms = [output_prefix + "0GenericAffine.mat"]
                    regFwdXfrms = [output_prefix + "0GenericAffine.mat"]
                else:
                    regInvXfrms = [output_prefix + "0GenericAffine.mat",
                                   output_prefix + "1InverseWarp.nii.gz"]                
                    regFwdXfrms = [output_prefix + "1Warp.nii.gz",
                                   output_prefix + "0GenericAffine.mat"]                
            else:
                multivariate_extras = list()
                for j in range(1, len(baseline_images)):
                    if "Quick" in type_of_transform:
                        multivariate_extras.append(["MI", baseline_images[j], followup_images[j], 1.0, 32])
                    else:
                        if "[s,2]" in type_of_transform or "[b,2,26]" in type_of_transform:
                            multivariate_extras.append(["CC", baseline_images[j], followup_images[j], 1.0, 2])
                        else:
                            multivariate_extras.append(["CC", baseline_images[j], followup_images[j], 1.0, 4])
                reg = ants.registration(baseline_images[0], followup_images[0], type_of_transform=type_of_transform, outprefix=output_prefix, multivariate_extras=multivariate_extras, verbose=1)
                if type_of_transform == "antsRegistrationSyNQuick[a]":
                    regInvXfrms = [output_prefix + "0GenericAffine.mat"]
                    regFwdXfrms = [output_prefix + "0GenericAffine.mat"]
                else:
                    regInvXfrms = [output_prefix + "0GenericAffine.mat",
                                   output_prefix + "1InverseWarp.nii.gz"]                
                    regFwdXfrms = [output_prefix + "1Warp.nii.gz",
                                   output_prefix + "0GenericAffine.mat"]                

            time_end = time.time()

            time_elapsed = time_end - time_start
            #######

            followup_landmark_filename = glob.glob(subjects[i] + "/" + subject_prefix + "_01_*_landmarks.csv")[0]
            moving_indices = pd.read_csv(followup_landmark_filename).drop('Landmark', axis=1)
            moving_indices = moving_indices.rename(columns={'X' : 'x', 'Y' : 'y', 'Z' : 'z'})
            moving_indices['y'] = moving_indices['y'] + 239

            moving_points = np.zeros(moving_indices.shape)
            for j in range(moving_indices.shape[0]):
                moving_points[j,:] = ants.transform_index_to_physical_point(baseline_images[0], (moving_indices.iloc[j].values).astype(int))

            moving_points_df = pd.DataFrame(data = {'x': moving_points[:,0], 'y': moving_points[:,1], 'z': moving_points[:,2]})
            if len(regInvXfrms) == 1:
                moving_warped_points = ants.apply_transforms_to_points(3, moving_points_df, regInvXfrms, whichtoinvert=(True,))   
            else:
                moving_warped_points = ants.apply_transforms_to_points(3, moving_points_df, regInvXfrms, whichtoinvert=(True, False))   
            moving_warped_points = moving_warped_points.to_numpy()   

            if write_xfrm_output:
                moving_warped_points_df = pd.DataFrame(data=moving_warped_points, columns=['X', 'Y', 'Z'])
                moving_warped_points_df.insert(0, "Landmark", list(range(1, moving_points.shape[0]+1)))
                moving_warped_points_df.to_csv(output_prefix + "moving_warped_landmarks.csv", index=False)

                warpedMovOut = ants.apply_transforms(baseline_images[0], followup_images[0], regFwdXfrms, whichtoinvert=(False, False))
                ants.image_write(reg['warpedmovout'], output_prefix + "moving_warped.nii.gz")

            baseline_landmark_filename = subjects[i] + "/" + subject_prefix + "_00_0000_landmarks.csv"
            if os.path.exists(baseline_landmark_filename):
                fixed_indices = pd.read_csv(baseline_landmark_filename).drop('Landmark', axis=1)
                fixed_indices = fixed_indices.rename(columns={'X' : 'x', 'Y' : 'y', 'Z' : 'z'})    
                fixed_indices['y'] = fixed_indices['y'] + 239

                fixed_points = np.zeros(fixed_indices.shape)
                for j in range(fixed_indices.shape[0]):
                    fixed_points[j,:] = ants.transform_index_to_physical_point(baseline_images[0], (fixed_indices.iloc[j].values).astype(int))

                average_distance_pre = 0.0
                average_distance_post = 0.0

                for j in range(fixed_points.shape[0]):
                    average_distance_pre += math.sqrt((fixed_points[j, 0] - moving_points[j, 0])**2 + 
                                                    (fixed_points[j, 1] - moving_points[j, 1])**2 +
                                                    (fixed_points[j, 2] - moving_points[j, 2])**2)
                    average_distance_post += math.sqrt((fixed_points[j, 0] - moving_warped_points[j, 0])**2 + 
                                                    (fixed_points[j, 1] - moving_warped_points[j, 1])**2 +
                                                    (fixed_points[j, 2] - moving_warped_points[j, 2])**2)


                average_distance_pre /= fixed_points.shape[0] 
                average_distance_post /= fixed_points.shape[0] 
                print("    Distance:  " + str(average_distance_pre) + " ---> " + str(average_distance_post))

                results_df.loc[len(results_df.index)] = [type_of_transform, '_'.join(modalities), subject_prefix, time_elapsed, average_distance_pre, average_distance_post]

                results_df.to_csv("/Users/ntustison/Desktop/resultsBrats.csv", index=False)

            


